### ISSUE TITLE
[JACK backend] supernova tells JACK server to not start if it's not already running

### ISSUE THREAD

**Issue by prko** [supercollider/supercollider#6148]
## Environment

* SuperCollider version: SuperCollider 3.14.0_dev
* Operating system: Ubuntu 22.04.3LTS (arm64)

## Steps to reproduce
Launch SuperCollider, then evaluate the following two-line code:
```
Server.supernova;
s.boot
```

## Expected vs. actual behavior
- expected: it should properly work.
- errors in Ubuntu 22.04
     ```
     -> localhost
     Booting server 'localhost' on address 127.0.0.1:57110.
     Supernova booting
      
     *** ERROR: could not start audio backend: Unable to connect to JACK server
     Server 'localhost' exited with exit code 0.
     ```

**Comment by Spacechild1** [comment]
The Windows issue is already known: https://github.com/supercollider/supercollider/issues/5899. I put this on my to-do list.

The Ubuntu issue is actually self explaining (`ERROR: could not start audio backend: Unable to connect to JACK server`). Is the Jack server running?

**Comment by prko** [comment]
> The Ubuntu issue is actually self explaining (`ERROR: could not start audio backend: Unable to connect to JACK server`). Is the Jack server running?

I think so, because server runs when using scsynth as a sound engine. So the following works:
```
Server.scsynth;
s.boot
```
However, the following does not work:
Server.supernova;
s.reboot
```
Should I do something after changing Server to supernova on Ubuntu?

**Comment by Spacechild1** [comment]
Can post the full console output, both for scsynth and supernova?

**Comment by prko** [comment]
I am attaching a video:
https://www.dropbox.com/scl/fi/itf4ymvrk56qcqx9qvfrb/supernova-and-scsynth-on-Ubuntu.mov?rlkey=h8yqsl1bw3bl545kzr68y04ob&dl=0

Summary: 
`Server.supernova` returns `-> Server`
`s.boot` returns the following error:
```
-> localhost
Booting server 'localhost' on address 127.0.0.1:57110.
Supernova booting

*** ERROR: could not start audio backend: Unable to connect to JACK server
Server 'localhost' exited with exit code 0.
```
`Server.scsynth;` returns `-> Server`
`s.boot` works as expected and returns the following messages on the Post window:
```
-> localhost
Booting server 'localhost' on address 127.0.0.1:57110.
jackdmp 1.9.20
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2021 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
audio_reservation_init
Acquire audio card Audio0
creating alsa driver ... hw:0|hw:0|1024|2|48000|0|0|nomon|swmeter|-|32bit
configuring for 48000Hz, period = 1024 frames (21.3 ms), buffer = 2 periods
ALSA: final selected sample format for capture: 32bit integer little-endian
ALSA: use 2 periods for capture
ALSA: final selected sample format for playback: 32bit integer little-endian
ALSA: use 2 periods for playback
JackDriver: client name is 'SuperCollider'
SC_AudioDriver: sample rate = 48000.000000, driver's block size = 1024
JackDriver: connected  system:capture_1 to SuperCollider:in_1
JackDriver: connected  system:capture_2 to SuperCollider:in_2
JackDriver: connected  SuperCollider:out_1 to system:playback_1
JackDriver: connected  SuperCollider:out_2 to system:playback_2
SuperCollider 3 server ready.
JackDriver: max output latency 42.7 ms
Requested notification messages from server 'localhost'
localhost: server process's maxLogins (1) matches with my options.
localhost: keeping clientID (0) as confirmed by server process.
Shared memory server interface initialized
server 'localhost' disconnected shared memory interface
Shared memory server interface initialized
```

**Comment by Spacechild1** [comment]
Ok, so the Jack server is definitely running and sclang can successfully connect to it, but Supernova can't. That's weird indeed.

Can you remove the Windows parts from the issue title and description, as it is already covered by https://github.com/supercollider/supercollider/issues/5899?

**Comment by dyfer** [comment]
@prko can you confirm that you are starting the JACK server before starting supernova? E.g. using QJackCtl or a similar tool? scsynth should start JACK automatically, I think.

I'm not familiar with this code, but [this possibly indicates that supernova is set to not start jack server automatically](https://github.com/supercollider/supercollider/blob/develop/server/supernova/audio_backend/jack_backend.hpp#L76C46-L76C53) if one is not already running. If that's the case, IMO this should be changed to match the scsynth behavior.

**Comment by prko** [comment]
I do not start Jack before booting scsynth and supernova. scsynth boots fine but supernova does not.

**Comment by prko** [comment]
@dyfer I do not start Jack before booting scsynth and supernova. scsynth boots fine but supernova does not.

**Comment by dyfer** [comment]
@prko please try booting JACK first and confirm whether supernova starts afterwards.

**Comment by prko** [comment]
@dyfer 

Supernova starts up after manually starting Jack, but `Jack: xrun detected - resyncing clock`  messages keep appearing:
```
Server.supernova
```
>-> Server

```
s.reboot
```
>-> localhost
Booting server 'localhost' on address 127.0.0.1:57110.
Supernova booting
Warning: cannot raise thread priority
Supernova ready
Warning: cannot raise thread priority
JackDriver: max output latency 42.7 ms
Jack: xrun detected - resyncing clock
Requested notification messages from server 'localhost'
localhost: no maxLogins info from server process.
localhost: keeping clientID (0) as confirmed by server process.
Shared memory server interface initialized
Jack: xrun detected - resyncing clock
Jack: xrun detected - resyncing clock
Jack: xrun detected - resyncing clock
Jack: xrun detected - resyncing clock
Jack: xrun detected - resyncing clock
Jack: xrun detected - resyncing clock

**Comment by dyfer** [comment]
In that case the issue is that supernova does not start jack automatically, not that it doesn't start. Do you mind if I change the title of the issue to reflect this?
Regarding xruns - this has to do with audio performance of your system and of supernova, which is possibly more demanding than scsynth. I'm not sure why it happens, but there might be multiple reasons... 

> Warning: cannot raise thread priority

This might be a clue, but I haven't looked into Linux audio for a while so I'm not sure how to proceed. Is your system optimized for audio? There's a ton of information about this online, but I'm not sure which one is still relevant, as it's been a tricky subject for decades...

**Comment by prko** [comment]
@dyfer 
>Do you mind if I change the title of the issue to reflect this?

I do not mind. I also see now that the title I gave is not appropriate to the phenomenon. You can change the title of the topic without asking me, as I am less experienced in this kind of work. To be honest, I seem to be collaborating with the other experts here, but in fact I am learning in a collaborative way, like an apprentice.

>Is your system optimized for audio? 

I do not know it. I have done nothing after installing Ubuntu. I only use Ubuntu to test the compatibility of the code I write. 
My Ubuntu is running under Parallels Desktop on a MacBookPro (M1max) and it must be slower than on the real hardware. However, you can usually tell by the dirty sound, which means that the sound is not properly processed in real time.

**Comment by JordanHendersonMusic** [comment]
On my system, scsynth does **not** start jack. So I wouldn't consider this a bug, at least a bug in supercollider itself, it might be an issue of jack configuration.

**Comment by dyfer** [comment]
@JordanHendersonMusic thanks for the extra info. Automatic starting of JACK depends on system configuration as well, though I don't remember specifics. 

However, looking at the code, this still seems to be an inconsistency between scsynth and supernova. When no JACK server name is specified, [scsynth opens JACK with `JackNullOption`](https://github.com/supercollider/supercollider/blob/develop/server/scsynth/SC_Jack.cpp#L237) passed to [JackOptions](https://jackaudio.org/api/types_8h.html#a396617de2ef101891c51346f408a375e), i.e. with default settings, while [supernova opens JACK with `JackNoStartServer` option](https://github.com/supercollider/supercollider/blob/develop/server/supernova/audio_backend/jack_backend.hpp#L76), explicitly requesting to not start JACK if it's not already running.

I think it would be reasonable to change the supernova to match scsynth in that regard, what do you think?

**Comment by JordanHendersonMusic** [comment]
Making these the same sounds good! I don't really understand what a 'null' option is, I wonder if supernova's implementation might be better as it is more explicit?

**Comment by dyfer** [comment]
My understanding is that "null" option simply means that no specific options are specified (see the [docs](https://jackaudio.org/api/types_8h.html#a396617de2ef101891c51346f408a375e) as what the options are). Not starting JACK is an explicit request in supernova the differs from scsynth's implementation, so I think it would be better to remove that (i.e. match scsynth).

**Comment by JordanHendersonMusic** [comment]
Just looking at the docs of the [function](https://jackaudio.org/api/group__ClientFunctions.html#gabbd2041bca191943b6ef29a991a131c5), only a [subset](https://jackaudio.org/api/types_8h.html#ab39e7f78dba631f3c0d34ac4cddcf80f ) can be used and null option isn't valid here.

**Comment by dyfer** [comment]
Hm, that's not how I'd read it... all the options are just various bits set to 1. `JackNullOption` seems to be simply no bits set (all zeroes - `0x00`), you can see this in the [source](https://jackaudio.org/api/types_8h_source.html) (line 107). I can't imagine how that would be invalid. 
In any case, since scsynth uses `JackNullOption` and it never came up as an issue, as far as I'm aware, I'd propose we change supernova to match this.

**Comment by prko** [comment]
I tested this using SuperCollider 3.15.0-dev built from detached 'HEAD' [6ec7382]:

---

```supercollider
Server.supernova;
```
returns:
```text
-> Server
```

---

```supercollider
s.boot
```
returns:
```text
-> localhost


*** Booting server 'localhost' on address 127.0.0.1:57110.
SC FFT global init: cosTable initialised.

*** Supernova booting
Warning: cannot raise thread priority
SuperCollider 3 server (supernova) ready.
Warning: cannot raise thread priority

Requested notification messages from server 'localhost'
localhost: no maxLogins info from server process.
localhost: keeping clientID (0) as confirmed by server process.
Shared memory server interface initialized
```

It seems to be working, so I am closing this issue. Thanks!

**Comment by ddw-js-70** [comment]
> It seems to be working

If I'm not mistaken, in PipeWire-enabled systems, if the PipeWire service is up and running, then PipeWire will respond to JACK connection requests. Supernova asks, "Is there a JACK service running?" PipeWire says, "Yep! Talk to me" and the server boot continues.

This is different from older Linux systems without PipeWire, where you're responsible for starting the JACK server yourself. Then, supernova asks, "Is there a JACK service running?" and... 🦗🦗🦗

Ubuntu 26.04 ships with PipeWire out of the box, so this issue may simply disappear as OS environments evolve.

**Comment by prko** [comment]
Oh, I tested on Ubuntu 24.04.4 LTS (ARM64) via Parallels Desktop on macOS. As I am not a daily Linux user, I was not aware of what you explained. Thanks!
